### DISCUSSION TITLE
How to remove drm from new books download by Kindle for Android (root required)

### DISCUSSION THREAD

**Discussion by oldonlytest** [noDRM/DeDRM_tools#440] [General]
I want to let you know If you have rooted android device you can actually remove drm from new purchased kfx books which you downloaded in your Android device. A must requrement for this is "ROOTED Android device" (Or rooted android emulator like Bluestacks or LDPlayer)

First you need DSN (Device Serial Number) from "Kindle for android"
For this use Root Browser / Explorer file managers and go to "/data/data/com.amazon.kindle" folder (can only do this in rooted android device/emulator
Open folder "shared_prefs"
Open or copy to internal storeage xml file named Notification.pref.xml
Open this xml file, you will find <string name="serializedDeviceAttributes"> line. In this line you will get 32 charactor long dsn number.
Example:        `<string name="serializedDeviceAttributes">{&quot;dsn&quot;:&quot;21da41a236a42fg2a9292a2cb24af43c&quot;,&quot;`

now convert this dsn to hexadesimal (online text to hex converter can do)  : 3231646134316132333661343266673261393239326132636232346166343363

Now create empty text file and type dsn number like this 
Example:      ` {"DSN": "3231646134316132333661343266673261393239326132636232346166343363"} `and save this file with extension *.k4i (kindleandroidkey.k4i)

Copy kindle book folder from android to pc (location Internal Storage / Android / data / com.amazon.kindle

Second step: Edit plugin to work with 32 charactor long DSN (Default for PC is 40 charactor)
Download plugin (Beta plugin 10.0.9)
https://github.com/noDRM/DeDRM_tools/releases/tag/v10.0.9

extract DeDRM_plugin.zip, find kfxdedrm.py file and open in notepad
edit this line 
`"for dsn_len,secret_len in [(0,0), (16,0), (16,40), (32,40), (40,0), (40,40)]:"`

to 

`for dsn_len,secret_len in [(0,0), (16,0), (16,40), (32,40), (40,0), (40,40), (32,0)]:`

so that plugin can accept 32 charator long dsn

Now zip again all files as DeDRM_plugin.zip 

Import plugin in "Calibre"

Import "kindleandroidkey.k4i" in DeDRM_tools (calibre/prefrence/plugin/dedrm/Customize Plugin/Configuration/Kindle for Mac/PC Ebooks/Import Existing Keyfile/

remove drm from new KFX books available in com.amazon.kindle /files

**Comment by ElleKayEm** [comment]
Instead of modifying the plugin to import the Android key file into the Kindle for PC/Mac configuration, couldn't you import it to the Kindle for Android configuration?

**Reply by oldonlytest** [reply]
Did not work for me, I think Android part of plugin is not updated to handle new kfx files

**Reply by ElleKayEm** [reply]
Ah, I see.  Makes sense.

**Comment by ghost** [comment]
Thanks for the info. I wrote a rough [bash script for automating key extraction](https://gist.github.com/neckothy/6c20a40c37c2d9af6ccc10f94f4c572d) from a rooted device if that is useful to anyone.

**Comment by mslmn** [comment]
> extract DeDRM_plugin.zip, find kfxdedrm.py file and open in notepad edit this line `"for dsn_len,secret_len in [(0,0), (16,0), (16,40), (32,40), (40,0), (40,40)]:"`

When I try to load the plugin into Calibre after editing kfxdedrm.py, I get the error `InvalidPlugin: The plugin 'DeDRM_plugin.zip' is invalid. It does not contain a top-level __init__.py file`

**Reply by oldonlytest** [reply]
v10.0.9
[DeDRM_plugin_androidkey.zip](https://github.com/noDRM/DeDRM_tools/files/13692936/DeDRM_plugin_androidkey.zip)
There should not be any subfolder named "DeDRM_plugin" in zip file

**Reply by 2023comicjtr** [reply]
What worked for me was opening the plugin zip file (rather than extracting the whole thing) in an archive manager (I used Peazip), then extracting only the kfxdedrm.py file in order to make the required change.  Once I saved the edited version of the file, I dragged it back into the still open plugin zip and allowed Peazip to resave it... I'd had the same error and that got rid of it for me.

**Comment by noDRM** [comment]
Interesting discovery. I've added that suggested code change to the kfxdedrm file so for newer versions (or if you're downloading the autorelease beta) that step is no longer needed.

**Comment by McSunnyDee** [comment]
Does this still work? I've got it to work for a good while, but sometime this month DeDRM started to fail decrypting the .ast files for my downloads.
I've checked my key and it seems to be the same.

**Reply by ElleKayEm** [reply]
I don't know about this working, but using an older version of Kindle for Android in BlueStacks is still working for me.

**Reply by oldonlytest** [reply]
Hi, can you share your downloaded book folder and "Notification.pref.xml"

**Reply by oldonlytest** [reply]
https://www.amazon.in/%E0%A4%85%E0%A4%95%E0%A4%AC%E0%A4%B0-%E0%A4%AC%E0%A5%80%E0%A4%B0%E0%A4%AC%E0%A4%B2-Hindi-PRAFULL-ACHARI-ebook/dp/B07VLVPSGQ

Hi, can you confirm whether this book is decryptable using old version of Kindle for Android?

**Reply by McSunnyDee** [reply]
Sorry, I'd rather not, now that I've got it back to working properly.
I can give some semi-useful info, though.

deDRM was failing to decrypt the voucher from the .ast file.

my "Notification.pref.xml" looks like you described and the books are from my personal library.

I've fixed it by doing a combination of things, but I have no idea what actually fixed it and I'm afraid to mess it up again.
It's very likely some, if not all, of these extra steps are completely useless.

- Downgraded the app to 8.100.1.100 (it's a very "crashy" version, don't really recommend it if it can be avoided) 

- Imported the android key from the "map_data_storage.db" file and imported the DSN from Notifications.pref.xml

- Finally, removed the deDRM plugin I had installed https://github.com/noDRM/DeDRM_tools/releases/tag/v10.0.9 and installed this one https://github.com/noDRM/DeDRM_tools_autorelease/releases after modifying the line from "kfxdedrm.py" to "dsn_len,secret_len in [(0,0), (16,0), (16,40), (32,40), (40,0), (40,40), (32,0), (22,0)]:"

- Restarted Calibre 7.8.

**Reply by JJSILDesigner** [reply]
Thank you _**McSunnyDee**_, I was racking my brain here. Before formatting the PC, the _**oldonlytest**_ method was working perfectly. Now I tested just importing the "map_data_storage.db" and nothing else, and it's working perfectly. There are files that work just by importing any .kfx file from the folder, others just by zipping everything and renaming it to kfz-zip. Let's see how long it will work.
